Add doc for List partition and List columns partition (#4522)
pingcap/docs: partitioned-table.md

* You want to use a column that contains time or date values, or containing values arising from some other series.
* You need to frequently run queries on the columns used for partitioning. For example, when executing a query like `EXPLAIN SELECT COUNT(*) FROM employees WHERE separated BETWEEN '2000-01-01' AND '2000-12-31' GROUP BY store_id;`, TiDB can quickly know that only the data in the `p2` partition needs to be scanned, because the other partitions do not match the `WHERE` condition.

### List partitioning

> **Warning:**
>
> List partitioning is an experimental feature introduced in v5.0.0-rc. It is not recommended that you use it in the production environment.

Before creating a List partitioned table, you need to set the value of the session variable `tidb_enable_table_partition` to `nightly`.

{{< copyable "sql" >}}

```sql
set @@session.tidb_enable_table_partition = nightly
```

> List partitioning is similar to Range partitioning. Unlike Range partitioning, in List partitioning, each partition contains rows for which the partitioning expression value belongs to a given data set. This data set defined for each partition has any number of values but cannot have duplicate values. You can use the `PARTITION ... VALUES IN (...)` clause to define these values.

Reviewer: @crazycs520 PTAL

> List partitioning is similar to Range partitioning. Unlike Range partitioning, in List partitioning, each partition contains rows for which the partitioning expression value belongs to a given data set. This data set defined for each partition has any number of values but cannot have duplicate values. You can use the `PARTITION ... VALUES IN (...)` clause to define these values.

Reviewer: ```suggestion
List partitioning is similar to Range partitioning. Unlike Range partitioning, in List partitioning, each partition contains rows for which the partitioning expression value belongs to a given value list. This value list defined for each partition has any number of values but cannot have duplicate values. You can use the `PARTITION ... VALUES IN (...)` clause to define these value lists.
```

> List partitioning is similar to Range partitioning. The main difference between them is that, in the List partitioning, for all rows in each partition of the table, the value computed by the partition expression belongs to the given data set. The data set defined by each partition has any number of values, but can not have duplicate values. You can use the `PARTITION ... VALUES IN (...)` clause to define these values.

Reviewer 2: ```suggestion
List partitioning is similar to Range partitioning. The main difference between them is that, in the List partitioning, for all rows in each partition of the table, the value computed by the partition expression belongs to the value lists that defined by each partition, and the value lists can not have duplicate values. You can use the `PARTITION ... VALUES IN (...)` clause to define these value lists.
```

Reviewer: The Chinese says "数据集". "value list" is what you mean, please also modify the Chinese version.

Reviewer 2: `value lists` can also translate to `数据集`, and `value lists` is more suitable to describe `List Partition`.

> Suppose that you want to create a personnel record form for a chain store. You can create a table as follows:

Reviewer: ```suggestion
Suppose that you want to create a personnel record table for a chain store. You can create a table as follows:
```

> Suppose you want to create a personnel record form. The example is as follows:

Reviewer 2: ```suggestion
Suppose you want to create a personnel record table. The example is as follows:
```

{{< copyable "sql" >}}

```sql
CREATE TABLE employees (
    id INT NOT NULL,
    hired DATE NOT NULL DEFAULT '1970-01-01',
    store_id INT
);
```

Suppose that there are 20 stores distributed in 4 districts, as shown in the table below:

```
| Region  | Store ID Numbers     |
| ------- | -------------------- |
| North   | 1, 2, 3, 4, 5        |
| East    | 6, 7, 8, 9, 10       |
| West    | 11, 12, 13, 14, 15   |
| Central | 16, 17, 18, 19, 20   |
```

> If you want to store the personnel data of employees of the same region in the same partition, you can create List partitions based on `store_id`:

Reviewer: ```suggestion
If you want to store the personnel data of employees of the same region in the same partition, you can create a List partitioned table based on `store_id`:
```

> If you want to store the personnel data of store employees of the same region in the same partition, you can create a List partitioning based on `store_id`:

Reviewer: ```suggestion
If you want to store the personnel data of employees of the same region in the same partition, you can create List partitions based on `store_id`:
```

> If you want to store the personnel data of store employees of the same region in the same partition, you can create a List partitioning based on `store_id`:

Reviewer 2: ```suggestion
If you want to store the personnel data of store employees of the same region in the same partition, you can create a List partition table based on `store_id`:
```

{{< copyable "sql" >}}

```sql
CREATE TABLE employees (
    id INT NOT NULL,
    hired DATE NOT NULL DEFAULT '1970-01-01',
    store_id INT
)
PARTITION BY LIST (store_id) (
    PARTITION pNorth VALUES IN (1, 2, 3, 4, 5),
    PARTITION pEast VALUES IN (6, 7, 8, 9, 10),
    PARTITION pWest VALUES IN (11, 12, 13, 14, 15),
    PARTITION pCentral VALUES IN (16, 17, 18, 19, 20)
);
```

> You can easily add or delete records related to a specific region in the table via the above statement. For example, suppose all the stores in the East region (East) are sold to another company, and all the row data related to the store employees in this region can be deleted by `ALTER TABLE employees TRUNCATE PARTITION pEast`, which can be executed much more efficiently than the equivalent `DELETE` statement `DELETE FROM employees WHERE store_id IN (6, 7, 8, 9, 10)`.

Reviewer: ```suggestion
After creating the partitions as above, you can easily add or delete records related to a specific region in the table. For example, suppose that all stores in the East region (East) are sold to another company. Then all the row data related to the store employees of this region can be deleted by executing `ALTER TABLE employees TRUNCATE PARTITION pEast`, which is much more efficient than the equivalent statement `DELETE FROM employees WHERE store_id IN (6, 7, 8, 9, 10)`.
```

> You can also use `ALTER TABLE employees DROP PARTITION pEast` to delete all these rows, but it also deletes the partition `pEast` from the definition of the table. Then you also need to use the `ALTER TABLE ... ADD PARTITION` statement to restore the original partitioning scheme of the table.

Reviewer: ```suggestion
You can also execute `ALTER TABLE employees DROP PARTITION pEast` to delete all related rows, but this statement also deletes the `pEast` partition from the table definition. In this situation, you must execute the `ALTER TABLE ... ADD PARTITION` statement to recover the original partitioning scheme of the table.
```

> Unlike Range partitioning, List partitioning does not have a similar `MAXVALUE` partition to store all values that do not belong to other partitions. Instead, all expected values of the partition expression must be included in the `PARTITION ... VALUES IN (...)` clause. If the value to be inserted in an `INSERT` statement does not match the column value of a partition, the statement fails to execute and an error is reported. See the following example:

Reviewer: ```suggestion
Unlike Range partitioning, List partitioning does not have a similar `MAXVALUE` partition to store all values that do not belong to other partitions. Instead, all expected values of the partition expression must be included in the `PARTITION ... VALUES IN (...)` clause. If the value to be inserted in an `INSERT` statement does not match the value of any partition, the statement fails to execute and an error is reported. See the following example:
```

> Unlike Range partitioning, the List partitioning does not have a similar `MAXVALUE` partitioning to store all values that do not belong to other partitions. All expected values of the partition expression should be included in the `PARTITION ... VALUES IN (...)` clause. If the value to be inserted by the `INSERT` statement does not match the column value of the partition, the statement fails to execute and reports an error, as shown in the following example:

Reviewer: ```suggestion
Unlike Range partitioning, List partitioning does not have a similar `MAXVALUE` partition to store all values that do not belong to other partitions. Instead, all expected values of the partition expression must be included in the `PARTITION ... VALUES IN (...)` clause. If the value to be inserted in an `INSERT` statement does not match the column value of a partition, the statement fails to execute and an error is reported. See the following example:
```

> Unlike Range partitioning, the List partitioning does not have a similar `MAXVALUE` partitioning to store all values that do not belong to other partitions. All expected values of the partition expression should be included in the `PARTITION ... VALUES IN (...)` clause. If the value to be inserted by the `INSERT` statement does not match the column value of the partition, the statement fails to execute and reports an error, as shown in the following example:

Reviewer 2: ```suggestion
Unlike Range partitioning, the List partitioning does not have a similar `MAXVALUE` partitioning to store all values that do not belong to other partitions. All expected values of the partition expression should be included in the `PARTITION ... VALUES IN (...)` clause. If the value to be inserted by the `INSERT` statement does not match the value of any partitions, the statement fails to execute and reports an error, as shown in the following example:
```

```sql
test> CREATE TABLE t (
    ->   a INT,
    ->   b INT
    -> )
    -> PARTITION BY LIST (a) (
    ->   PARTITION p0 VALUES IN (1, 2, 3),
    ->   PARTITION p1 VALUES IN (4, 5, 6)
    -> );
Query OK, 0 rows affected (0.11 sec)

test> INSERT INTO t VALUES (7, 7);
ERROR 1525 (HY000): Table has no partition for value 7
```

> To ignore the above types of errors, you can use the `IGNORE` keyword. After using this keyword, you avoid inserting rows containing unmatched partition column values but insert any rows with matching values, and no errors are reported:

Reviewer: ```suggestion
To ignore the error type above, you can use the `IGNORE` keyword. After using this keyword, if a row contains values that are not matched the partition's column values, this row will not be inserted. Instead, any row with matched values is inserted, and no error is reported:
```

> To ignore the above types of errors, you can use the `IGNORE` keyword. After using this keyword, you avoid inserting rows containing unmatched partition column values but insert any rows with matching values, and no errors are reported:

Reviewer 2: ```suggestion
To ignore the above types of errors, you can use the `IGNORE` keyword. After using this keyword, you avoid inserting rows containing unmatched partition value lists but insert any rows with matching values, and no errors are reported:
```

> To ignore the error type above, you can use the `IGNORE` keyword. After using this keyword, if a row contains values that do not match the partition's column values, this row will not be inserted. Instead, any row with matched values is inserted, and no error is reported:

Reviewer: ```suggestion
To ignore the error type above, you can use the `IGNORE` keyword. After using this keyword, if a row contains values that do not match the partition's value lists, this row will not be inserted. Instead, any row with matched values is inserted, and no error is reported:
```

```sql
test> TRUNCATE t;
Query OK, 1 row affected (0.00 sec)

test> INSERT IGNORE INTO t VALUES (1, 1), (7, 7), (8, 8), (3, 3), (5, 5);
Query OK, 3 rows affected, 2 warnings (0.01 sec)
Records: 5  Duplicates: 2  Warnings: 2

test> select * from t;
+------+------+
| a    | b    |
+------+------+
|    5 |    5 |
|    1 |    1 |
|    3 |    3 |
+------+------+
3 rows in set (0.01 sec)
```

### List COLUMNS partitioning

> **Warning:**
>

> > This feature is introduced in v5.0.0-rc. It is currently an experimental feature, and it is not recommended that you use it in the production environment.

Reviewer: ```suggestion
> List COLUMNS partitioning is an experimental feature introduced in v5.0.0-rc. It is not recommended that you use it in the production environment.
```

> List COLUMNS partitioning is a variant of List partitioning. You can use multiple columns as partition keys, and columns of data types other than integer types as partition columns. You can also use string type, `DATE` and `DATETIME` columns.

Reviewer: ```suggestion
List COLUMNS partitioning is a variant of List partitioning. You can use multiple columns as partition keys and use columns of the data types other than integer as partition columns. You can also use columns of the string, `DATE`, and `DATETIME` types.
```

> Suppose you want to divide the store employees of the following 12 cities into 4 regions according to relevant regulations, as shown in the following table:

Reviewer: ```suggestion
Suppose that you want to divide the store employees from the following 12 cities into 4 regions, as shown in the following table:
```

```
| Region | Cities                         |
| :----- | ------------------------------ |
| 1      | LosAngeles,Seattle, Houston    |
| 2      | Chicago, Columbus, Boston      |
| 3      | NewYork, LongIsland, Baltimore |
| 4      | Atlanta, Raleigh, Cincinnati   |
```

> You can use List Column partitioning to create a table for employee data, and store each row of data in the partition corresponding to the employee's city, as shown below:

Reviewer: ```suggestion
You can use List COLUMNS partitioning to create a table and store each row in the partition that corresponds to the employee's city, as shown below:
```

{{< copyable "sql" >}}

```sql
CREATE TABLE employees_1 (
    id INT NOT NULL,
    fname VARCHAR(30),
    lname VARCHAR(30),
    hired DATE NOT NULL DEFAULT '1970-01-01',
    separated DATE NOT NULL DEFAULT '9999-12-31',
    job_code INT,
    store_id INT,
    city VARCHAR(15)
)
PARTITION BY LIST COLUMNS(city) (
    PARTITION pRegion_1 VALUES IN('LosAngeles', 'Seattle', 'Houston'),
    PARTITION pRegion_2 VALUES IN('Chicago', 'Columbus', 'Boston'),
    PARTITION pRegion_3 VALUES IN('NewYork', 'LongIsland', 'Baltimore'),
    PARTITION pRegion_4 VALUES IN('Atlanta', 'Raleigh', 'Cincinnati')
);
```

> Unlike the List partitioning, you don't need to use the expression in the `COLUMNS()` clause to convert column values to integers.

Reviewer: ```suggestion
Unlike List partitioning, in List COLUMNS partitioning, you do not need to use the expression in the `COLUMNS()` clause to convert column values to integers.
```

> List COLUMNS partitioning can also be partitioned using columns of type `DATE` and `DATETIME`, as shown in the following example. This example uses the same names and columns as the previous `employees_1` table, but uses List COLUMNS partitioning based on the `hired` column:

Reviewer: ```suggestion
List COLUMNS partitioning can also be implemented using columns of the `DATE` and `DATETIME` types, as shown in the following example. This example uses the same names and columns as the previous `employees_1` table, but uses List COLUMNS partitioning based on the `hired` column:
```

{{< copyable "sql" >}}

```sql
CREATE TABLE employees_2 (
    id INT NOT NULL,
    fname VARCHAR(30),
    lname VARCHAR(30),
    hired DATE NOT NULL DEFAULT '1970-01-01',
    separated DATE NOT NULL DEFAULT '9999-12-31',
    job_code INT,
    store_id INT,
    city VARCHAR(15)
)
PARTITION BY LIST COLUMNS(hired) (
    PARTITION pWeek_1 VALUES IN('2020-02-01', '2020-02-02', '2020-02-03',
        '2020-02-04', '2020-02-05', '2020-02-06', '2020-02-07'),
    PARTITION pWeek_2 VALUES IN('2020-02-08', '2020-02-09', '2020-02-10',
        '2020-02-11', '2020-02-12', '2020-02-13', '2020-02-14'),
    PARTITION pWeek_3 VALUES IN('2020-02-15', '2020-02-16', '2020-02-17',
        '2020-02-18', '2020-02-19', '2020-02-20', '2020-02-21'),
    PARTITION pWeek_4 VALUES IN('2020-02-22', '2020-02-23', '2020-02-24',
        '2020-02-25', '2020-02-26', '2020-02-27', '2020-02-28')
);
```

> Besides, you can also add multiple columns in the `COLUMNS()` clause, for example:

Reviewer: ```suggestion
In addition, you can also add multiple columns in the `COLUMNS()` clause. For example:
```

### Compatibility with MySQL

> Currently, TiDB only supports Range partitioning, List partitioning, List COLUMNS partitioning and Hash partitioning. Other partitioning types that are available in MySQL such as key partitioning are not supported yet in TiDB.

Reviewer: ```suggestion
Currently, TiDB only supports Range partitioning, List partitioning, List COLUMNS partitioning, and Hash partitioning. Other partitioning types that are available in MySQL such as key partitioning are not supported yet in TiDB.
```
